content/post/ddos-detection.md

```diff
@@ -0,0 +1,413 @@
+
+```python
+class ServerPartitionFunction(object):
+    def __init__(self, partitions):
+        self.part_size = 2**32/partitions
+
+    def partition(self, data):
+        parts = data['server'].split('.')
+        ipnum = ((int(parts[0]) << 24) + (int(parts[1]) << 16) +
+                 (int(parts[2]) << 8) + int(parts[3]))
+        return ipnum/self.part_size
+```
+
+The state object is where things get interesting. For each server, we need to maintain a 60 seconds long rolling window of aggregate data. This data is small enough that we could simply use a list and pay the cost of shifting it whenever we need to roll the window. But a much more efficient way of doing this, especially if your data window is larger, is to use a [circular buffer]. For brevity, the code for the circular buffer is not included in this post, but you may find it alongside the [demo code][ddos-detection].
+
+Each server's state needs to keep track of a few things:
+
+1. The server's address
+2. rolling window of aggregate per-second data for requests per second and unique clients per second
+3. The weighted mean of the requests data over the window
+4. The weighted standard deviation of the requests data over the window
+5. The weighted mean of the clients data over the window
+6. The weighted standard deviation of the clients data over the window
+7. The current partial second's timestamp
+8. The number of requests for the current (partial) second
+9. A set of unique client addresses (to produce the number of clients per second aggregate)
+10. The current health status of the server
+
+```python
+class SingleServerLoadStatistics(object):
+    def __init__(self, address):
+        self.address = address
+        self.window = CircularBuffer(60)
+        self.req_mu = 0
+        self.req_sigma = 0
+        self.cli_mu = 0
+        self.cli_sigma = 0
+        self.current_ts = 0
+        self.current_req = 0
+        self.current_cli = set()
+        self.is_attack = False
+```
+
+That seems simple enough. But we also need to implement three pieces of logic:
+
+1. Advancing the window and recomputing the means and standard deviations
+2. Updating the current (partial) second's data
+3. Determine whether the server is under attack
+
+The weighted mean is given by ![weighted mean][weighted mean formula]
+while the weighted standard deviation is given by ![weighted sigma][weighted sigma formula]
+You can find the code for both in the `stats.py` file in the [demo code][ddos-detection].
```

Comment:
I dont think I get anything from the includes weighted mean and weighted sigma functions. they are distracting. my first thought was "o look a formula. i dont care enough about this post to invest time in figuring out what it is doing, i'll just skip that" and when I skip it, i generally skip over other info. i think its more than enough to say "hey we get the weighted mean" and leave it at that. 

this level of detail is too much in my mind.  you are very in the weeds here. the important part of this is the streaming and wallaroo aspect of things.

Reply:
@SeanTAllen I have some questions
1. Is your general idea here that the post should say "we compute a weighted means and standard deviation" without any further detail (equations or code)?
2. Or is that we should say the above, but add a single sentence with link to code to those interested, e.g. "and if you're interested in how, check out the [code here][...]"
3. Or that maybe we _should_ include the equations, but maybe either in an appendix at the bottom, or perhaps in the readme for the code, and link to that?


My original thinking was that it would be nice to show something non-trivial—equations and all—being implemented with Wallaroo.

Reply:
I ended up removing this part and leaving a direct reference to the stats.py file in the code repo.